freshmaker/handlers/botas/botas_shipped_advisory.py

```diff
@@ -270,24 +306,112 @@ def handle(self, event):
 
-        self.start_to_build_images(builds)
-        msg = f"Advisory {db_event.search_key}: Rebuilding " \
-              f"{len(db_event.builds.all())} bundle images."
-        db_event.transition(EventState.BUILDING, msg)
+        :param db_event: database event that represent rebuild event
+        :rtype: list
+        :return: list of advisories that should be rebuilt
+        """
+        default_bundle_data = {
+            'nvr': '',
+            'update': {},
+            'append': {},
+            'pullspecs': [],
+        }
+        # map with old pullspec and bundle info
+        old_pullspec_to_bundle_map = dict()
+        # get all replaced pullspecs from provided 'bundle_images'
+        for bundle_nvr in self.event.bundle_images:
+            artifact_build = db.session.query(ArtifactBuild).filter(
+                ArtifactBuild.rebuilt_nvr == bundle_nvr,
+                ArtifactBuild.type == ArtifactType.IMAGE.value,
+            ).one_or_none()
+            if artifact_build is None:
+                log.warning(f'Can\'t find build for a bundle image "{bundle_nvr}"')
+                continue
+            pullspec_overrides = artifact_build.bundle_pullspec_overrides
+            for pullspec in pullspec_overrides:
+                old_pullspec = pullspec.get('_old', None)
+                if old_pullspec is None:
+                    continue
+                old_pullspec_to_bundle_map.setdefault(
+                    old_pullspec, copy.deepcopy(default_bundle_data))
+                # bundle will now contain new pullspec overrides
+                old_pullspec_to_bundle_map[old_pullspec]['pullspecs'] = pullspec_overrides
+                break
+
+        if not old_pullspec_to_bundle_map:
+            msg = 'None of the bundle images have applicable pullspecs to replace'
+            log.warning(msg)
+            db_event.transition(EventState.SKIPPED, msg)
+            return []
 
-        return []
+        # Unauthenticated koji session to fetch build info of bundles
+        koji_api = KojiService(conf.koji_profile)
+        rebuild_nvr_to_bundle_map = dict()
+        # compare replaced pullspecs with pullspecs in 'container_images' and
+        # create map for bundles that should be rebuilt with their nvrs
+        for container_image_nvr in self.container_images:
+            # Fetch buildinfo
+            buildinfo = koji_api.get_build(container_image_nvr)
+            # Get the original pullspecs
+            pullspecs = (
+                buildinfo.get('extra', {})
+                         .get('image', {})
+                         .get('operator_manifests', {})
+                         .get('related_images', {})
+                         .get('pullspecs', [])
+            )
+            for pullspec in pullspecs:
+                if pullspec.get('new') not in old_pullspec_to_bundle_map:
+                    continue
+                bundle = old_pullspec_to_bundle_map[old_pullspec]
```

Comment:
So we don't want to replace all the pullspecs, we just want to replace the pullspecs in `container_images` that still reference `_old` in any of `bundle_images`.

So the flow would be as follows:
* Create a dictionary of `_old` to `new` pullspec entries
* Loop through each entry in `bundle_images` and for each entry, loop through their pullspecs stored in FM if available
* For each pullspec that has `_old`, create an entry in the dictionary of `_old` to `new`. This gives you the all possible replacements that could apply to any of the entries in `container_images`.
* For each entry in `container_images`, get the pullspecs from Freshmaker if present, if not, fallback to Brew for this metadata
* For each pullspec in the pullspecs of an entry in `container_images`, check if the `new` pullspec has an entry in the `_old` to `new` dictionary, if yes, replace `new` with this entry in the dictionary.

Does this make sense?

Reply:
Yes, I will do it this way.